### PR TITLE
Make ClientEvents, CollectorFilter and Message types generic.

### PR DESCRIPTION
**Pull request by VimHax**
**Please describe the changes this PR makes and why it should be merged:**

I've added some generic type parameters for the following types for the respective reasons stated below.

-> `CollectorFilter` - I feel that this type would benefit from being able to define exactly what parameters will be passed in to the function. Not only does this leave less room for mistakes but it also makes it so users don't have to work with the type `any` as much when they're using TypeScript.
-> `ClientEvents` - This type has been given a generic parameter which dictates which events will emit partial data. This is useful as redundant checking for partial data won't be required if the user has not enabled the respective partial in the `ClientOptions`.
-> `Message` - The generic parameter for this type describes whether or not the message was sent in a guild. When you're working with messages you often have to check whether or not `Message#guild` is null even when you're pretty certain it isn't, inserting `Message#guild as Guild` everywhere is also much less elegant. This parameter can also be used to add more type information to the return types of some functions, such as `Message#reply` which will always return a `Message` which was sent in a guild if the original message was sent in a guild.

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Comment by iCrawl** [comment]
This needs a rebase.

**Comment by VimHax** [comment]
I merged the changes into master, will that do?

File: typings/index.d.ts [hunk 1]
```diff
@@ -958,15 +958,15 @@ declare module 'discord.js' {
     public toString(): string;
   }
 
-  export class Message extends Base {
-    constructor(client: Client, data: object, channel: TextChannel | DMChannel | NewsChannel);
+  export class Message<G extends boolean = boolean> extends Base {
```
Comment (NotSugden) [inline typings/index.d.ts:961]: maybe could rename the type parameter to like `Guild` or smth, G feels like a weird name

Comment (VimHax) [reply typings/index.d.ts:961]: But wouldn't that clash with the `Guild` type?

Comment (NotSugden) [reply typings/index.d.ts:961]: true, the point was to have a better name for it tho

Comment (VimHax) [reply typings/index.d.ts]: I made the name `SentInGuild` and changed some other names too... Hopefully that has improved the situation.

**Comment by VimHax** [comment]
Should've made separate PRs for each of the features I implemented.

**Pull request closed without merging**